src/app_main.c

```diff
@@ -108,23 +121,103 @@ void app_init() {
     USB_power(0);
     USB_power(1);
     view_idle(0);
+
+    // set the default bip32 path
+    bip32_depth = 5;
+    uint32_t new_bip32_path[] = {
+         44 | 0x80000000,  // purpose
+        714 | 0x80000000,  // coin type (chain ID)
+          0 | 0x80000000,  // account
+          0,               // change (no change addresses for now)
+          0,               // address index
+    };
+    memcpy(bip32_path, new_bip32_path, sizeof(bip32_path));
 }
 
-bool extractBip32(uint8_t *depth, uint32_t path[10], uint32_t rx, uint32_t offset) {
+// extract_bip32 extracts the bip32 path from the apdu buffer
+bool extract_bip32(uint8_t *depth, uint32_t path[5], uint32_t rx, uint32_t offset) {
     if (rx < offset + 1) {
         return 0;
     }
 
     *depth = G_io_apdu_buffer[offset];
     const uint16_t req_offset = 4 * *depth + 1 + offset;
 
-    if (rx < req_offset || *depth > 10) {
+    if (rx < req_offset || *depth != 5) {
         return 0;
     }
     memcpy(path, G_io_apdu_buffer + offset + 1, *depth * 4);
     return 1;
 }
 
+// validate_bnc_bip32 checks the given bip32 path against an expected one
+bool validate_bnc_bip32(uint8_t depth, uint32_t path[5]) {  // path is 10 bytes for compatibility
+    // Only paths in the form 44'/714'/{account}'/0/{index} are supported
+    // placeholder for a user-specified value is 0xFFFFFFFF
+    uint32_t placeholder = 0xFFFFFFFF;
+    uint32_t expected[] = {
+         44 | 0x80000000,  // purpose
+        714 | 0x80000000,  // coin type (chain ID)
+          0 | 0x80000000,  // account
+          0,               // change (no change addresses for now)
+          placeholder,     // address index
+    };
+    if (depth != 5) {
+        return 0;
+    }
+    if (sizeof(expected) / 4 != depth) {
+        return 0;
+    }
+    for (unsigned int i = 0; i < depth; i++) {
+        if (expected[i] == placeholder) continue;
+        if (path[i] != expected[i]) return 0;
+    }
+    return 1;
+}
+
+void set_hrp(char *hrp) {
+    strcpy(bech32_hrp, hrp);
+    bech32_hrp_len = strlen(bech32_hrp);
+}
```

Comment:
Maybe we should check the string length of hrp against `MAX_BECH32_HRP_LEN`

Reply:
This is only used with [hardcoded string consts](https://github.com/binance-chain/ledger-app-binance/blob/d1aebe6240759ae2c8dc75c366ed92612bdd652e/src/view.c#L106)`"bnb"` and `"tbnb"`.

`extract_hrp` and `validate_bnc_hrp` are used for user provided hrp validation.